File: Single Cycle Check/hasSingleCycle.js
function hasSingleCycle(array) {
  let elementsVisited = 0;
  let startingIdx = 0;
  while (elementsVisited < array.length) {
    if (elementsVisited > 0 && startingIdx === 0) {
      return false;
    }

    elementsVisited++;
    startingIdx = nextIdx(startingIdx, array);
  }

  return startingIdx === 0;
}

function nextIdx(startingIdx, array) {
  const jump = array[startingIdx];
  const nextIdx = (startingIdx + jump) % array.length;
  return nextIdx >= 0 ? nextIdx : nextIdx + array.length;
}

exports.hasSingleCycle = hasSingleCycle;

File: Single Cycle Check/hasSingleCycle.cpp
#include <vector>
using namespace std;

int nextIdx(int startingIdx, vector<int> array);

bool hasSingleCycle(vector<int> array)
{
    int elementsVisited = 0;
    int startingIdx = 0;
    while (elementsVisited < array.size())
    {
        if (elementsVisited > 0 && startingIdx == 0)
        {
            return false;
        }
        elementsVisited++;
        startingIdx = nextIdx(startingIdx, array);
    }
    return startingIdx == 0;
}

int nextIdx(int startingIdx, vector<int> array)
{
    int jump = array[startingIdx];
    int nextIdx = (startingIdx + jump) % (int)array.size();
    return nextIdx >= 0 ? nextIdx : nextIdx + array.size();
}